model/src/main/kotlin/config/FileListingStorageConfiguration.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2023 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.model.config
+
+import org.apache.logging.log4j.kotlin.Logging
+
+import org.ossreviewtoolkit.model.utils.DatabaseUtils
+import org.ossreviewtoolkit.model.utils.FileProvenanceFileStorage
+import org.ossreviewtoolkit.model.utils.PostgresProvenanceFileStorage
+import org.ossreviewtoolkit.model.utils.ProvenanceFileStorage
+import org.ossreviewtoolkit.utils.ort.ortDataDirectory
+import org.ossreviewtoolkit.utils.ort.storage.FileStorage
+import org.ossreviewtoolkit.utils.ort.storage.LocalFileStorage
+
+private const val TABLE_NAME = "provenance_file_listings"
+private const val FILENAME = "file_listings.xz"
+
+data class FileListingStorageConfiguration(
+    /**
+     * Configuration of the [FileStorage] used for storing the file listings.
+     */
+    val fileStorage: FileStorageConfiguration? = null,
+
+    /**
+     * Configuration of the [PostgresProvenanceFileStorage] used for storing the file listings.
+     */
+    val postgresStorage: PostgresStorageConfiguration? = null
+) {
+    private companion object : Logging
+
+    fun createStorage(): ProvenanceFileStorage =
+        when {
+            fileStorage != null -> {
+                if (postgresStorage != null) {
```

Comment:
Personally, I'd prefer a more "direct" check for this in an `init`-block like done in `FileArchiverConfiguration`.

Reply:
In general, I wonder whether this should resemble `FileArchiverConfiguration` more, also regarding the extension function for the creation, and the general `enable` flag. I'm curious about @mnonnenmacher's opinion here.

Reply:
> Personally, I'd prefer a more "direct" check for this in an `init`-block like done in `FileArchiverConfiguration`.

That mentioned init block is executed when ort config is loaded. I thought it would be better to only log it exactly when the storage gets created. I still think it's no a bad idea, but I do not have such a strong opinion on that, so I could move it to `init` if you want.

Reply:
> In general, I wonder whether this should resemble `FileArchiverConfiguration` more, also regarding the extension function for the creation, and the general `enable` flag. I'm curious about @mnonnenmacher's opinion here.

It would have been nice to hear some rationale for this. Anyhow, I deliberately have made it mandatory so that we can count on the file listings to be present for simplicity. Making the file listing optional I prefer to not be part of the (initial) implementation.

Reply:
> In general, I wonder whether this should resemble FileArchiverConfiguration more, also regarding the extension function for the creation, and the general enable flag. I'm curious about @mnonnenmacher's opinion here.

I need to have a look at the code tomorrow before I can answer that.

> Anyhow, I deliberately have made it mandatory so that we can count on the file listings to be present for simplicity. Making the file listing optional I prefer to not be part of the (initial) implementation.

IMO the file lists should be mandatory, this makes it easier to build the features that depend on it.

> One does not have to, but should configure a storage backend in ~/.ort/config/config.yml. If one doesn't do so, ORT falls back to storing the file listings under ~/.ort which may not be desired.

Off topic, but this reminds me of #5516 which would prevent that we have yet another place to configure the same database connection. I wonder if we really need the flexibility to configure all storage types for the scanner separately, or if one central configuration to configure either a file storage or a Postgres storage would be good enough.

Reply:
@mnonnenmacher I could add #5516 to my epic, e.g. as last clean-up ticket. Ok?

Reply:
> > In general, I wonder whether this should resemble `FileArchiverConfiguration` more, also regarding the extension function for the creation, and the general `enable` flag. I'm curious about @mnonnenmacher's opinion here.
> 
> It would have been nice to hear some rationale for this.

IMO it's wrong that `FileListingStorageConfiguration` knows how to create `ProvenanceFileStorage`. It should be the opposite: `ProvenanceFileStorage` knowns how create itself given a `FileListingStorageConfiguration`. However, the latter is hard to implement cleanly as we can have either a `FileProvenanceFileStorage` or a `PostgresProvenanceFileStorage`. So we need to move creation to a "higher level" / common place, which leads us back to `FileListingStorageConfiguration`. But *if* we do it in `FileListingStorageConfiguration`, I find it to be more clean if it's separated from the data class via an extension function like done in `FileArchiverConfiguration`.

Reply:
> > Anyhow, I deliberately have made it mandatory so that we can count on the file listings to be present for simplicity. Making the file listing optional I prefer to not be part of the (initial) implementation.
> 
> IMO the file lists should be mandatory, this makes it easier to build the features that depend on it.

Why is it okay to have an optional file storage for license files, but it's not okay to make the storage for listings optional?

If the argument is simplicity, we should also make the license file storage mandatory, as anyway pretty much everybody seems to agree that the original wording of the license files needs to be retained.

Reply:
> But if we do it in FileListingStorageConfiguration, I find it to be more clean if it's separated from the data class

I'm not following the arguments "separation" and "more clean, to me it's not obvious  

1. why an extension function is more clean
2. in what way the degree of separation is increased.

I'll change it anyway

Reply:
> > we should also make the license file storage mandatory, as anyway pretty much everybody seems to agree that the original wording of the license files needs to be retained.
> 
> Maybe, but can we leave this discussion out of scope here?

Sorry, no. I'm not saying we must already implement this now, but we should be allowed to discuss it, as it obviously has an impact on this PR.

I have a problem with bending arguments and changing minds every time it suites somebody's needs, just because there is pressure to implement a feature. Out of a sudden "simplicity" is an argument here, whereas in the past we were forced to add complexity by hiding several (IMO uncritical) features behind feature flags. I just miss the consistency and equality in the line of though here.

Reply:
> I'm not following the arguments "separation" and "more clean, to me it's not obvious
> 
>     1. why an extension function is more clean
>     2. in what way the degree of separation is increased.

My answer to both is: Because the implementation is not part of the class body anymore. (I.e. it's more of a "visual" separation that expresses the intent better, as that's probably the best we can do here.)